src/forum/forum_post.rs

```diff
@@ -162,9 +136,58 @@ impl ForumPost {
             parent_id: Some(parent_id),
             author,
             created_at,
+            deleted_at: None,
             message,
         };
 
         Ok(forum_reply)
     }
+
+    pub fn soft_delete_post(conn: &rusqlite::Connection, id: usize) -> ForumResult<()> {
+        let deleted_at = chrono::Local::now().naive_local();
+        conn.execute(
+            "UPDATE forum_posts SET deleted_at = ?1 WHERE id = ?2",
+            (deleted_at, id),
+        )
+        .map_err(ForumError::DatabaseError)?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct PostTreeNode {
+    pub post: ForumPost,
+    pub replies: Vec<PostTreeNode>,
+}
+
+impl PostTreeNode {
+    pub fn build_tree(conn: &rusqlite::Connection, parent_id: usize) -> ForumResult<Vec<Self>> {
+        let mut stmt = conn
+            .prepare(
+                "
+            SELECT id, root_id, parent_id, created_at, deleted_at, author, message
+            FROM forum_posts
+            WHERE parent_id = ?1
+            ORDER BY created_at ASC
+            ",
+            )
+            .map_err(ForumError::DatabaseError)?;
+
+        let mut nodes = stmt
+            .query_map([parent_id], |row| {
+                Ok(PostTreeNode {
+                    post: ForumPost::get_from_db(row).unwrap(),
+                    replies: Vec::new(),
+                })
+            })
+            .map_err(ForumError::DatabaseError)?
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(ForumError::DatabaseError)?;
+
+        for node in &mut nodes {
+            node.replies = Self::build_tree(conn, node.post.id)?;
+        }
+
+        Ok(nodes)
+    }
```

Comment:
Awesome!

I detect an [N+1 SQL](https://stackoverflow.com/questions/97197/what-is-the-n1-selects-problem-in-orm-object-relational-mapping) problem here: Each call performs a database query. The idea behind the "root_id" database field is so that you can build an entire thread from the results of a single SQL query.

If/when you wish to render only a sub-thread, you can't as efficiently select all the nodes for any sub-tree, only the root. So for now, if you are drawing only the tree under a given parent, you can fetch all posts for their common root_id, and disregard the ones that are not children of the given parent. (They will still arrive in the right order.) (Other optimisations can apply in a later PR.)